### ISSUE TITLE
bug when saving annotations over LAN: timeout

### ISSUE THREAD

**Issue by LukeAI** [openvinotoolkit/cvat#2886]
### My actions before raising this issue
- [X] Read/searched [the docs](https://github.com/opencv/cvat/tree/master#documentation)
- [X] Searched [past issues](/issues)

We are getting an odd persistent bug with a local CVAT install when trying to save annotations. This was present in CVAT 1.2 and 1.0

CVAT is installed on a local server, when accessed across the Internet, no problems are experienced. When accessed over the LAN, sometimes when saving progress - the save will lag indefinitely with "Created Objects are being saved on the server", the save will fail and annotation progress will be lost.

In the browser console:

```
DevTools failed to load SourceMap: Could not load content for http://MY_DOMAIN:MY_PORT/download.worker.e2283bafc6215c9d8a80.js.map: Load canceled due to load timeout+
```

(MY_DOMAIN and MY_PORT are our externally facing IP address and port) set in override.yml like:
```
services:
  cvat_proxy:
    environment:
      CVAT_HOST: MY_DOMAIN
    ports:
      - "MY_PORT:80"
```

In `docker logs cvat` I see:

```
2021-03-01 12:01:51,644 DEBG 'rqworker_default_0' stderr output:
DEBUG:rq.worker:Sent heartbeat to prevent worker timeout. Next one should arrive within 480 seconds.

2021-03-01 12:01:51,644 DEBG 'rqworker_default_1' stderr output:
DEBUG:rq.worker:Sent heartbeat to prevent worker timeout. Next one should arrive within 480 seconds.

2021-03-01 12:01:51,644 DEBG 'rqworker_low' stderr output:
DEBUG:rq.worker:Sent heartbeat to prevent worker timeout. Next one should arrive within 480 seconds.

2021-03-01 12:01:51,644 DEBG 'rqworker_low' stderr output:
INFO:rq.worker:Cleaning registries for queue: low

```

The LAN is plenty fast enough

### Your Environment

- Git hash commit (`git log -1`):  checked out tag 1.2
- Docker version `docker version` (e.g. Docker 17.0.05): 20.10.4
- Are you using Docker Swarm or Kubernetes? No
- Operating System and version (e.g. Linux, Windows, MacOS): ubuntu 16.04

I realise that diagnosing and fixing this bug which does not even happen in a predictable way on my end may be very challenging. but perhaps even some general guidance on how to avoid or workaround would be wonderful.

**Comment by azhavoro** [comment]
Did you tried to save the annotation one more time? Also please obtain the full error message of the failed request in the browser console, I suppose it will be 50x error code.
If you specified CVAT_HOST: MY_DOMAIN, you can access your cvat host over the Internet regardless of whether clients are on the same local network or not. Please provide details about the failed save annotation request.

**Comment by LukeAI** [comment]
If I try to save again (after refreshing in the browser) it will often work. This error just occurs randomly - quite a few annotators have complained about it running a few different OS's - I advise everybody to use Chrome (is that still a recommendation?) some have reported firefox to be more stable but may be a coincidence.

What I have written here is the full error message. yes, typing MY_DOMAIN:MY_PORT works on the LAN or over the internet, but this problem has only been reported on the LAN.

**Comment by azhavoro** [comment]
Try to save the annotation several times without reloading the browser tab.


> What I have written here is the full error message.

When this error occurs, open browser developer tools and go to the Console tab, you should receive an error message about a failed attempt to save the annotations.

**Comment by LukeAI** [comment]
![error](https://user-images.githubusercontent.com/43993778/109781097-8600ac00-7bff-11eb-94b5-4a55baa837b1.png)
This is the whole error message, as it appears in the console browser. I'm told that it is less likely to occur in firefox and more likely to occur when there are more than a few annotations to save at a time.

**Comment by LukeAI** [comment]
![error](https://user-images.githubusercontent.com/43993778/109782117-6fa72000-7c00-11eb-9fcb-e772972e44e5.png)
This is the whole error message, as it appears in the browser console.
I am told that it seems to be more likely to occur in chrome, and more likely to occur, the more unsaved annotations there are.

**Comment by azhavoro** [comment]
This warning is not related to the error you are talking about.

**Comment by LukeAI** [comment]
hmm ok. I did clear the console before the error happened, and that is all there was, after it happened.

**Comment by nmanovic** [comment]
@LukeAI , need logs from your containers.

**Comment by LukeAI** [comment]
@nmanovic which logs other than `docker logs cvat` ? all the other dockers?

**Comment by nmanovic** [comment]
@LukeAI , let's start with docker logs cvat. To be honest I don't what we are looking for. I hope to see some anomalies.

**Comment by nmanovic** [comment]
@LukeAI , I can see only a couple of lines of the log.

**Comment by LukeAI** [comment]
@nmanovic I don't want to attach the entire log here due to concerns about leaking compromising information but I have emailed it directly to you.

**Comment by LukeAI** [comment]
@nmanovic  was that any use?

**Comment by nmanovic** [comment]
@LukeAI , we need a clear way to reproduce the problem on our side. If we cannot, we will not be able to fix. I thought logs could answer on some questions, but you could share them.

 You need to understand that the problem can be in your LAN. Thus if my team doesn't have a way to reproduce, you need to investigate on your own.

We have an internal CVAT instance in LAN and don't see any problems with saving annotations.

**Comment by LukeAI** [comment]
hmm ok, I understand, thanks anyway.

**Comment by LukeAI** [comment]
I just got a report of the issue:

docker logs cvat:
```
2021-03-18 10:47:57,620 DEBG 'rqworker_low' stderr output:
DEBUG:rq.worker:Sent heartbeat to prevent worker timeout. Next one should arrive within 480 seconds.

2021-03-18 10:49:57,585 DEBG 'rqworker_default_1' stderr output:
DEBUG:rq.worker:Sent heartbeat to prevent worker timeout. Next one should arrive within 480 seconds.

2021-03-18 10:49:57,885 DEBG 'rqworker_default_0' stderr output:
DEBUG:rq.worker:Sent heartbeat to prevent worker timeout. Next one should arrive within 480 seconds.

2021-03-18 10:50:10,701 DEBG 'runserver' stderr output:
[Thu Mar 18 10:50:10.700968 2021] [wsgi:error] [pid 1098:tid 140016575829760] [remote 172.28.0.6:53536] INFO:cvat.client:{"client_id":610103,"name":"Send user activity","time":"2021-03-18T10:50:10.642000Z","payload":{"working_time":0},"is_active":true,"username":"sam-rahimi"}

2021-03-18 10:51:07,628 DEBG 'runserver' stderr output:
[Thu Mar 18 10:51:07.628375 2021] [wsgi:error] [pid 1098:tid 140016601007872] [remote 172.28.0.6:53578] INFO:cvat.client.task_2198:{"job_id":2176,"task_id":2198,"client_id":610103,"name":"Load job","time":"2021-03-18T10:50:10.643000Z","payload":{"duration":945,"frame count":32,"track count":0,"object count":0,"box count":0,"polygon count":0,"polyline count":0,"points count":0,"cuboids count":0,"tag count":0},"is_active":true,"username":"sam-rahimi"}

2021-03-18 10:51:07,629 DEBG 'runserver' stderr output:
[Thu Mar 18 10:51:07.628685 2021] [wsgi:error] [pid 1098:tid 140016601007872] [remote 172.28.0.6:53578] INFO:cvat.client.task_2198:{"job_id":2176,"task_id":2198,"client_id":610103,"name":"Fit image","time":"2021-03-18T10:50:11.595000Z","payload":{},"is_active":true,"username":"sam-rahimi"}
[Thu Mar 18 10:51:07.628905 2021] [wsgi:error] [pid 1098:tid 140016601007872] [remote 172.28.0.6:53578] INFO:cvat.client.task_2198:{"job_id":2176,"task_id":2198,"client_id":610103,"name":"Zoom image","time":"2021-03-18T10:50:20.047000Z","payload":{},"is_active":true,"username":"sam-rahimi"}
[Thu Mar 18 10:51:07.629116 2021] [wsgi:error] [pid 1098:tid 140016601007872] [remote 172.28.0.6:53578] INFO:cvat.client.task_2198:{"job_id":2176,"task_id":2198,"client_id":610103,"name":"Zoom image","time":"2021-03-18T10:50:21.163000Z","payload":{},"is_active":true,"username":"sam-rahimi"}

2021-03-18 10:51:07,629 DEBG 'runserver' stderr output:
[Thu Mar 18 10:51:07.629315 2021] [wsgi:error] [pid 1098:tid 140016601007872] [remote 172.28.0.6:53578] INFO:cvat.client.task_2198:{"job_id":2176,"task_id":2198,"client_id":610103,"name":"Zoom image","time":"2021-03-18T10:50:25.462000Z","payload":{},"is_active":true,"username":"sam-rahimi"}

2021-03-18 10:51:07,629 DEBG 'runserver' stderr output:
[Thu Mar 18 10:51:07.629516 2021] [wsgi:error] [pid 1098:tid 140016601007872] [remote 172.28.0.6:53578] INFO:cvat.client.task_2198:{"job_id":2176,"task_id":2198,"client_id":610103,"name":"Draw object","time":"2021-03-18T10:50:32.176000Z","payload":{"count":1,"duration":13041},"is_active":true,"username":"sam-rahimi"}
[Thu Mar 18 10:51:07.629747 2021] [wsgi:error] [pid 1098:tid 140016601007872] [remote 172.28.0.6:53578] INFO:cvat.client.task_2198:{"job_id":2176,"task_id":2198,"client_id":610103,"name":"Change frame","time":"2021-03-18T10:50:33.606000Z","payload":{"from":0,"to":1},"is_active":true,"username":"sam-rahimi"}

2021-03-18 10:51:07,630 DEBG 'runserver' stderr output:
[Thu Mar 18 10:51:07.629924 2021] [wsgi:error] [pid 1098:tid 140016601007872] [remote 172.28.0.6:53578] INFO:cvat.client.task_2198:{"job_id":2176,"task_id":2198,"client_id":610103,"name":"Fit image","time":"2021-03-18T10:50:33.637000Z","payload":{},"is_active":true,"username":"sam-rahimi"}
[Thu Mar 18 10:51:07.630097 2021] [wsgi:error] [pid 1098:tid 140016601007872] [remote 172.28.0.6:53578] INFO:cvat.client.task_2198:{"job_id":2176,"task_id":2198,"client_id":610103,"name":"Zoom image","time":"2021-03-18T10:50:40.039000Z","payload":{},"is_active":true,"username":"sam-rahimi"}

2021-03-18 10:51:07,630 DEBG 'runserver' stderr output:
[Thu Mar 18 10:51:07.630271 2021] [wsgi:error] [pid 1098:tid 140016601007872] [remote 172.28.0.6:53578] INFO:cvat.client.task_2198:{"job_id":2176,"task_id":2198,"client_id":610103,"name":"Change frame","time":"2021-03-18T10:50:40.718000Z","payload":{"from":1,"to":0},"is_active":true,"username":"sam-rahimi"}
[Thu Mar 18 10:51:07.630444 2021] [wsgi:error] [pid 1098:tid 140016601007872] [remote 172.28.0.6:53578] INFO:cvat.client.task_2198:{"job_id":2176,"task_id":2198,"client_id":610103,"name":"Change frame","time":"2021-03-18T10:50:41.209000Z","payload":{"from":0,"to":1},"is_active":true,"username":"sam-rahimi"}

2021-03-18 10:51:07,630 DEBG 'runserver' stderr output:
[Thu Mar 18 10:51:07.630617 2021] [wsgi:error] [pid 1098:tid 140016601007872] [remote 172.28.0.6:53578] INFO:cvat.client.task_2198:{"job_id":2176,"task_id":2198,"client_id":610103,"name":"Change frame","time":"2021-03-18T10:50:43.540000Z","payload":{"from":1,"to":0},"is_active":true,"username":"sam-rahimi"}

2021-03-18 10:51:07,631 DEBG 'runserver' stderr output:
[Thu Mar 18 10:51:07.630790 2021] [wsgi:error] [pid 1098:tid 140016601007872] [remote 172.28.0.6:53578] INFO:cvat.client.task_2198:{"job_id":2176,"task_id":2198,"client_id":610103,"name":"Change frame","time":"2021-03-18T10:50:45.413000Z","payload":{"from":0,"to":1},"is_active":true,"username":"sam-rahimi"}
[Thu Mar 18 10:51:07.630963 2021] [wsgi:error] [pid 1098:tid 140016601007872] [remote 172.28.0.6:53578] INFO:cvat.client.task_2198:{"job_id":2176,"task_id":2198,"client_id":610103,"name":"Drag object","time":"2021-03-18T10:50:49.880000Z","payload":{"id":1},"is_active":true,"username":"sam-rahimi"}

2021-03-18 10:51:07,631 DEBG 'runserver' stderr output:
[Thu Mar 18 10:51:07.631135 2021] [wsgi:error] [pid 1098:tid 140016601007872] [remote 172.28.0.6:53578] INFO:cvat.client.task_2198:{"job_id":2176,"task_id":2198,"client_id":610103,"name":"Resize object","time":"2021-03-18T10:50:53.713000Z","payload":{"id":1},"is_active":true,"username":"sam-rahimi"}
[Thu Mar 18 10:51:07.631307 2021] [wsgi:error] [pid 1098:tid 140016601007872] [remote 172.28.0.6:53578] INFO:cvat.client.task_2198:{"job_id":2176,"task_id":2198,"client_id":610103,"name":"Change frame","time":"2021-03-18T10:50:53.803000Z","payload":{"from":1,"to":2},"is_active":true,"username":"sam-rahimi"}

2021-03-18 10:51:07,631 DEBG 'runserver' stderr output:
[Thu Mar 18 10:51:07.631479 2021] [wsgi:error] [pid 1098:tid 140016601007872] [remote 172.28.0.6:53578] INFO:cvat.client.task_2198:{"job_id":2176,"task_id":2198,"client_id":610103,"name":"Drag object","time":"2021-03-18T10:50:55.872000Z","payload":{"id":1},"is_active":true,"username":"sam-rahimi"}
[Thu Mar 18 10:51:07.631651 2021] [wsgi:error] [pid 1098:tid 140016601007872] [remote 172.28.0.6:53578] INFO:cvat.client.task_2198:{"job_id":2176,"task_id":2198,"client_id":610103,"name":"Resize object","time":"2021-03-18T10:50:59.184000Z","payload":{"id":1},"is_active":true,"username":"sam-rahimi"}

2021-03-18 10:51:07,632 DEBG 'runserver' stderr output:
[Thu Mar 18 10:51:07.631824 2021] [wsgi:error] [pid 1098:tid 140016601007872] [remote 172.28.0.6:53578] INFO:cvat.client.task_2198:{"job_id":2176,"task_id":2198,"client_id":610103,"name":"Resize object","time":"2021-03-18T10:51:01.518000Z","payload":{"id":1},"is_active":true,"username":"sam-rahimi"}

2021-03-18 10:51:07,632 DEBG 'runserver' stderr output:
[Thu Mar 18 10:51:07.631997 2021] [wsgi:error] [pid 1098:tid 140016601007872] [remote 172.28.0.6:53578] INFO:cvat.client.task_2198:{"job_id":2176,"task_id":2198,"client_id":610103,"name":"Change frame","time":"2021-03-18T10:51:01.920000Z","payload":{"from":2,"to":3},"is_active":true,"username":"sam-rahimi"}
[Thu Mar 18 10:51:07.632170 2021] [wsgi:error] [pid 1098:tid 140016601007872] [remote 172.28.0.6:53578] INFO:cvat.client.task_2198:{"job_id":2176,"task_id":2198,"client_id":610103,"name":"Save job","time":"2021-03-18T10:51:06.443000Z","payload":{"duration":1141},"is_active":true,"username":"sam-rahimi"}

2021-03-18 10:51:07,632 DEBG 'runserver' stderr output:
[Thu Mar 18 10:51:07.632358 2021] [wsgi:error] [pid 1098:tid 140016601007872] [remote 172.28.0.6:53578] INFO:cvat.client.task_2198:{"job_id":2176,"task_id":2198,"client_id":610103,"name":"Send task info","time":"2021-03-18T10:51:07.586000Z","payload":{"frame count":32,"track count":1,"object count":3,"box count":1,"polygon count":0,"polyline count":0,"points count":0,"cuboids count":0,"tag count":0},"is_active":true,"username":"sam-rahimi"}
[Thu Mar 18 10:51:07.632554 2021] [wsgi:error] [pid 1098:tid 140016601007872] [remote 172.28.0.6:53578] INFO:cvat.client.task_2198:{"job_id":2176,"task_id":2198,"client_id":610103,"name":"Send user activity","time":"2021-03-18T10:51:07.587000Z","payload":{"working_time":56880},"is_active":true,"username":"sam-rahimi"}
```

Browser console
```
TypeError: this.contentWindow is null PictureInPictureChild.jsm:569:5

InvalidStateError: JSWindowActorChild.document getter: Cannot access property 'document' after actor 'PictureInPictureToggle' has been destroyed PictureInPictureChild.jsm:225
Content Security Policy: Ignoring “'unsafe-inline'” within script-src: ‘strict-dynamic’ specified
Content Security Policy: Ignoring “https:” within script-src: ‘strict-dynamic’ specified
Content Security Policy: Ignoring “http:” within script-src: ‘strict-dynamic’ specified
debuggee 'resource://devtools/shared/base-loader.js:289' would run
```

don't know if this helps?

**Comment by LukeAI** [comment]
@nmanovic ok I believe this is resolved / worked-around, I added the local 192.168.X IP address to HOST_NAMES and when annotators access CVAT using the local IP, the problem does not come up.

It's a little strange that it works most of the time, but not always, when using the Internet-facing domain name from within the LAN, but it's not really a problem, now that I know how to workaround.
